src/index.ts

```diff
@@ -53,17 +63,371 @@ export function fromWrappedInstanceMethod<
   };
 }
 
-export function WithNoopPreAuth() {
-  return WithPreAuth(() => true);
+export function NoopPreAuth() {
+  return PreAuthorize(() => true);
 }
 
 export function NoopFinalAuth() {
-  return WithFinalAuth(() => true);
+  return FinalAuthorize(() => true);
+}
+
```

Comment:
- [ ] Are you sure that the way projector is written, you won't lose type information?
- [ ] Don't we need to do some type constraint stuff so that typescript will complain if the incoming context is missing a member on whereToLook?
- [ ] Make all these so that InitPreContext(projector) isn't called EVERY request... In other words, do that before the return statement.

Reply:
actually about the third bullet above, I don't think this is right - it probably won't play well with the rest of the library...  And it's not consistent...  {preAuthorize: blah} ... blah should always be a function that expects the full context.  Right now it won't be...  Instead you need to turn it inside out and do something like this...
```
return InitPreContext((htCtx: sometype) => projector(htCtx[whereToLook]));
```
and
```
return InitPreContext((htCtx: sometype) => { return {[whereToStore]: projector(htCtx[whereToLook])};});
```
- [ ] inside out

Reply:
it's all done but we will have a problem with that
the reason is the functions like regular AttachData and others breaks automatic type guessing
basically to make it work correctly functions should now the full type of all context so it will restore automatic type guessing
and i didn't figure out how to do that for now

Reply:
so even regular AttachData() breaks the type guessing?
I see it's explicitly typing projector to be a function that takes any and returns any.
What if you make projector generic (and type constrain it to be of the right shape)?

The mappings of the specific peculiarities of each projector are starting to get repetitive, so it's probably a good idea to factor them out and reuse them...
```
export interface DoWorkProjector = (context: TContextIn) => PromiseOrSync<TContextOut>

export interface OptionallyHasDoWork<TContextIn, TContextOut> {
  doWork?: DoWorkProjector<TContextIn, TContextOut>;
}
```
As a bonus, it'll also help to do this more cleanly
https://github.com/captaincaius/hipthrusts/blob/master/src/types.ts#L65

- [ ] loss of typing

Reply:
as discussed, in THIS PR let's fix the loss of typing everywhere, but postpone the fix for rxjs'ish auto-type-guessing

Reply:
done
i don't do that refactor stuff you post above for now
i got error with not assignable types for all this functions